lib/http-context.js

```diff
@@ -374,15 +405,97 @@ function toXML(input) {
   }
   return xml;
 }
+
+HttpContext.prototype.shouldReturnEventStream = function() {
+  var req = this.req;
+  var query = req.query;
+  var format = query._format;
+
+  var acceptable = req.accepts('text/event-stream');
+  return (format === 'event-stream') || acceptable;
+};
+
+HttpContext.prototype.respondWithEventStream = function(stream) {
+  var client = new SSEClient(this.req, this.res);
+
+  client.initialize();
+
+  stream.on('data', function(chunk) {
+    client.send('data', JSON.stringify(chunk));
+  });
+
+  stream.on('error', function(err) {
+    var outErr = {message: err.message};
+    for (var key in err) {
+      outErr[key] = err[key];
+    }
+    client.send('error', JSON.stringify(outErr));
+  });
+
+  stream.on('end', function() {
+    client.send({event: 'end', data: 'null'});
+  });
+};
+
 /**
  * Finish the request and send the correct response.
  */
 
-HttpContext.prototype.done = function() {
+HttpContext.prototype.done = function(cb) {
+  var ctx = this;
+  var method = this.method;
+  var streamsDesc = method.streams;
+  var returnStreamDesc = streamsDesc && streamsDesc.returns;
+  var methodReturnsStream = !!returnStreamDesc;
+  var res = this.res;
+  var out = this.io && this.io.out;
+  var err = this.error;
+  var result = this.result;
+
+  if (methodReturnsStream) {
+    if (returnStreamDesc.json) {
+      debug('handling json stream');
+
+      res.setHeader('Content-Type',
+      returnStreamDesc.contentType || 'application/json; boundary=NL');
+      res.setHeader('Transfer-Encoding', 'chunked');
+
+      var stream = result[returnStreamDesc.arg];
+
+      if (returnStreamDesc.type === 'ReadableStream') {
+        if (ctx.shouldReturnEventStream()) {
+          debug('respondWithEventStream');
+          ctx.respondWithEventStream(stream);
+          return;
+        }
+
+        debug('piping to mdm stream');
+        stream.pipe(out);
+        stream.on('error', function(err) {
+          var outErr = {message: err.message};
+          for (var key in err) {
+            outErr[key] = err[key];
+          }
+
+          // this is the reason we are using mux-demux
+          out.error(outErr);
+
+          out.end();
+        });
+        // TODO(ritch) support multi-part streams
+      } else {
+        cb(new Error('unsupported stream type: ' + returnStreamDesc.type));
+      }
+    } else {
+      cb(new Error('Unsupported stream descriptor, only descriptors ' +
+        'with property "json:true" are supported'));
+    }
```

Comment:
Nitpick: I would personally move the contents of the block `if(methodReturnsStream)` to a new method, e.g. `HttpContext.prototype._streamResponse`. The purpose is to keep the `done` method reasonably short.